bson/marshal.go

```diff
@@ -162,8 +170,9 @@ func MarshalAppendWithRegistry(r *bsoncodec.Registry, dst []byte, val interface{
 //
 // See [Encoder] for more examples.
 func MarshalAppendWithContext(ec bsoncodec.EncodeContext, dst []byte, val interface{}) ([]byte, error) {
-	sw := new(bsonrw.SliceWriter)
-	*sw = dst
+	sw := bufPool.Get().(*bytes.Buffer)
+	defer bufPool.Put(sw)
```

Comment:
We should check the size and utilization of of `sw` and only recycle it if it's not too large and not underutilized.

E.g.
```go
defer func() {
	if sw.Cap() < 16*1024*1024 && sw.Cap()/2 < sw.Len() {
		bufPool.Put(sw)
	}
}()
```

See the similar code in `Operation.Execute` for details:
https://github.com/mongodb/mongo-go-driver/blob/99d53d6348c0e7dce87255dce18b1f6660717dec/x/mongo/driver/operation.go#L519-L531